File: common/src/main/java/com.abc/adapter/VendorResponseAdapter.java
package com.abc.adapter;

import com.abc.model.VendorResponse;
import com.google.gson.*;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class VendorResponseAdapter implements JsonSerializer<VendorResponse[]>, JsonDeserializer<VendorResponse[]> {

    @Override
    public VendorResponse[] deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        List<VendorResponse> res = new ArrayList<>();
        for (JsonElement arrayElement : json.getAsJsonArray()){
            try {
                JsonObject jsonObj = arrayElement.getAsJsonObject();
                String type = jsonObj.get("type").getAsString();
                JsonElement element = jsonObj.get("props");
                res.add(context.deserialize(element, Class.forName("com.abc.model.vendor."+type)));
            } catch (ClassNotFoundException e) {
                System.out.println("add logs");
            }
        }
        return res.toArray(new VendorResponse[res.size()]);
    }

    @Override
    public JsonElement serialize(VendorResponse[] src, Type typeOfSrc, JsonSerializationContext context) {
        JsonArray res = new JsonArray();
        for (VendorResponse resp : src) {
            JsonObject element = new JsonObject();
            element.add("type", new JsonPrimitive(resp.getClass().getSimpleName()));
            element.add("props", context.serialize(resp, resp.getClass()));
            res.add(element);
        }
        return res;
    }
}

File: common/src/main/java/com.abc/adapter/VendorResponseAdapter.kt
package com.abc.adapter

import com.abc.model.VendorResponse
import com.google.gson.*
import java.lang.reflect.Type
import java.util.*

class VendorResponseAdapter : JsonSerializer<Array<VendorResponse>>, JsonDeserializer<Array<VendorResponse>> {

    override fun serialize(src: Array<VendorResponse>, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val res = JsonArray();
        for (resp in src) {
            val element = JsonObject()
            element.add("type", JsonPrimitive(resp.javaClass.simpleName))
            element.add("props", context.serialize(resp, resp.javaClass))
            res.add(element)
        }
        return res
    }

    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Array<VendorResponse> {
        return json.asJsonArray.mapNotNullTo(ArrayList<VendorResponse>()) {
            try {
                val jsonObj = it.asJsonObject
                val type = jsonObj["type"].asString
                val element = jsonObj["props"]
                context.deserialize(element, Class.forName("com.abc.model.vendor.$type"))
            } catch (e: ClassNotFoundException) {
                null
            }
        }.toTypedArray()
    }

}